### PR TITLE
nixos/postgresql: enrich ensure users and databases support

### PR DESCRIPTION
**Pull request by Scrumplex**
## Description of changes

Based on #203474

## TODOs

- [x] Update release notes
- [x] Add tests for `ensureClauses`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/tests/hedgedoc.nix [hunk 1]
```diff
@@ -33,10 +33,19 @@ import ./make-test-python.nix ({ pkgs, lib, ... }:
...
+              owner = "hedgedoc";
+            }
+          ];
+          ensureUsers = [
+            {
+              name = "hedgedoc";
+              passwordFile = pkgs.writeText "hedgedoc-password" "snakeoilpassword";
+              ensureClauses.login = true;
+            }
+          ];
```
Comment (SuperSandro2000) [inline nixos/tests/hedgedoc.nix:48]: I don't know why the dbname and owner don't match here and why not socket auth is used but I also don't have enough time to also maintain the tests of my packages. sigh. :(

File: nixos/modules/services/databases/postgresql.nix [hunk 2]
```diff
@@ -532,7 +622,67 @@ in
...
+              ++ addS "LC_CTYPE" ps.lcCtype
+              ++ addL "ALLOW_CONNECTIONS" ps.allowConnections
+              ++ addL "CONNECTION LIMIT" ps.connectionLimit
+              ++ addL "IS_TEMPLATE" ps.isTemplate);
+
+          # Run a single SQL command using psql with a list of additional
+          # command line options.
+          psqlCmd = cmd: args: concatStringsSep " " (
+            [ "$PSQL -tA" ]
+            ++ args
```
Comment (marsam) [inline nixos/modules/services/databases/postgresql.nix:645]: `args` doesn't seem to be used. What do you think of removing it?

**Review by marsam — APPROVED** [review]
This looks great. Thanks a lot!

**Comment by marsam** [comment]
@GrahamcOfBorg test postgresql

File: nixos/modules/services/databases/postgresql.nix (hunk 2 above)
Comment (Scrumplex) [reply nixos/modules/services/databases/postgresql.nix:646]: `args` is actually used a few lines below in `ensureUsers`

**Comment by bendlas** [comment]
Unfortunately, It looks like this will a bit late for helping us adapt to postgres 15 for 23.11. For now, we'll deprecate `ensurePermissions` and instead advise for `ensureDBOwnership` + the occasional raw `postStart` script in the manual.

I would very much like to give a proper look at this approach for 24.05, especially also look at ways this could tie into system - wide approach towards state management and migrations. Also I'd like to look for a way to get the benefits of ensure-style options, while avoiding the pitfalls of [convergent configuration management](https://github.com/NixOS/nixpkgs/issues/206467#issuecomment-1806336858)

**Comment by Scrumplex** [comment]
I'll close this as I don't really intend on working on this much further. Though feel free to continue this, as we don't have an alternative to convergent configuration management for databases in Nixpkgs atm.

**Pull request closed without merging**